src/routes/v2_reroute.rs

```diff
@@ -0,0 +1,146 @@
+use super::v3::project_creation::CreateError;
+use crate::{
+    database::{
+        models::{version_item, DatabaseError},
+        redis::RedisPool,
+    },
+    models::ids::VersionId,
+    util::actix::{generate_multipart, MultipartSegment, MultipartSegmentData},
+};
+use actix_multipart::Multipart;
+use actix_web::http::header::{HeaderMap, TryIntoHeaderPair};
+use actix_web::HttpResponse;
+use futures::{stream, StreamExt};
+use serde_json::{json, Value};
+
+pub async fn set_side_types_from_versions<'a, E>(
+    json: &mut serde_json::Value,
+    exec: E,
+    redis: &RedisPool,
+) -> Result<(), DatabaseError>
+where
+    E: sqlx::Executor<'a, Database = sqlx::Postgres>,
+{
+    json["client_side"] = json!("required"); // default to required
+    json["server_side"] = json!("required");
+    let version_id = json["versions"].as_array().and_then(|a| a.iter().next());
+    if let Some(version_id) = version_id {
+        let version_id = serde_json::from_value::<VersionId>(version_id.clone())?;
+        let versions_item = version_item::Version::get(version_id.into(), exec, redis).await?;
+        if let Some(versions_item) = versions_item {
+            json["client_side"] = versions_item
+                .version_fields
+                .iter()
+                .find(|f| f.field_name == "client_side")
+                .map(|f| f.value.serialize_internal())
+                .unwrap_or(json!("required"));
+            json["server_side"] = versions_item
+                .version_fields
+                .iter()
+                .find(|f| f.field_name == "server_side")
+                .map(|f| f.value.serialize_internal())
+                .unwrap_or(json!("server_side"));
+        }
+    }
+    Ok(())
+}
+
+// TODO: this is not an ideal way to do this, but it works for now
+pub async fn extract_ok_json(response: HttpResponse) -> Result<serde_json::Value, HttpResponse> {
+    if response.status() == actix_web::http::StatusCode::OK {
+        let failure_http_response = || {
+            HttpResponse::InternalServerError().json(json!({
+                "error": "reroute_error",
+                "description": "Could not parse response from V2 redirection of route."
+            }))
+        };
+        // Takes json out of HttpResponse, mutates it, then regenerates the HttpResponse
+        let body = response.into_body();
+        let bytes = actix_web::body::to_bytes(body)
+            .await
+            .map_err(|_| failure_http_response())?;
+        let json_value: Value =
+            serde_json::from_slice(&bytes).map_err(|_| failure_http_response())?;
+        Ok(json_value)
+    } else {
+        Err(response)
+    }
+}
+
+pub async fn alter_actix_multipart(
```

Comment:
Worth considering, not sure if it's worth doing - the methods in this file seem like great candidates for more detailed unit-level tests just around them, no integration. Seems like there are probably a lot of edge cases to be tested here, which can be done faster and more precisely than at the integration level

Reply:
Yeah definitely.  I think I want to come back to write a lot of tests around this area as I refactor it a little bit. Tbh Im not sure how I feel about the different ways to handle these rerouting functions so far, none of the ideas Ive run through have really felt 100% right